Add area of applicability calculations

mikemahoney218 opened the pull request:
This PR is a first stab at adding functions for calculating the [area of applicability](https://besjournals.onlinelibrary.wiley.com/doi/10.1111/2041-210X.13650) of a model. 

I patterned everything here off of `apd_pca`, so might have accidentally copied something without realizing that it only applies to that function. 

``` r
library(vip)
#> 
#> Attaching package: 'vip'
#> The following object is masked from 'package:utils':
#> 
#>     vi
library(applicable)
#> Loading required package: ggplot2
train <- gen_friedman(1000, seed = 101)
test <- train[701:1000, ]
train <- train[1:700, ]

pp <- ppr(y ~ ., data = train, nterms = 11)
importance <- vi_permute(
  pp,
  target = "y",
  metric = "rsquared",
  pred_wrapper = predict
)

apd_aoa(y ~ ., train, test, importance = importance)
#> # Predictors:
#>    10
#> Area-of-applicability threshold:
#>    0.0890452

apd_aoa(y ~ ., train, test, importance = importance) |> 
  score(test)
#> # A tibble: 300 × 2
#>         di aoa  
#>      <dbl> <lgl>
#>  1 0.0230  TRUE 
#>  2 0.00907 TRUE 
#>  3 0.0126  TRUE 
#>  4 0.0121  TRUE 
#>  5 0.129   FALSE
#>  6 0.0261  TRUE 
#>  7 0.0154  TRUE 
#>  8 0.00707 TRUE 
#>  9 0.141   FALSE
#> 10 0.0159  TRUE 
#> # … with 290 more rows
```

<sup>Created on 2022-06-27 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

mikemahoney218 commented:
I tagged everyone who mentioned wanting to be kept looped-in on this for a review -- let me know if you want me to not tag you going forward!

topepo commented:
I'll do a more in-depth review soon but I think that using "area of applicability" as the method name is redundant and not specific enough. 

I would suggest `apd_dissimilarity_index`, `apd_di`, or something similar.

mikemahoney218 commented:
Renamed to `apd_di` in d7f83b5 .

topepo commented on R/di-fit.R:
```diff
@@ -0,0 +1,276 @@
+#' this method focused on spatial models, there is nothing inherently spatial
+#' about the method; it can be used with any type of data.
+#'
+#' The `importance` argument is structured to work with objects returned by the
```
We should define some sort of tidy format consistent with broom (at least in the naming convention). I think columns  `term` and `estimate` (or maybe `value`) would be good. 

Also, we need to differentiate between a zero importance due to scaling (like caret does) or due to not being used in the model. Also also, we might want to require all of the variables to be included. 

Can you define a straw man specification for that?

topepo replied:
Maybe I missed it but is there any validation code for the concordance of the values in the importance df and the columns of the data being passed? 

For example, how would dummy variables be dealt with between the two objects, etc?

topepo replied:
I'm not sure that we can do anything about this but what happens if there is a difference in data processing between the model and the APD method? So, let's say a model uses a recipe to preprocess/transform the data. If the recipe doesn't match that, will that cause an error or poor results? 

It would be good to say something about that in the documentation (as a reminder).

topepo commented on R/di-fit.R:
```diff
@@ -0,0 +1,276 @@
+#'
+#'   * A __data frame__ containing the predictors.
+#'
+#'  If `validation` is `NULL`, then this function will use within-sample distances.
```
Can you give information here of what is being validated? Are there specific metrics being computed? What is the downside of not using a validation set etc.

mikemahoney218 replied:
I tried to clarify this in 86b1a5de37230c2aca4daef69ae3de936f02dc79 and 33faf2e but am not sure if I quite nailed it -- it should be the same hold-out set as you used in calculating all other model metrics. Please let me know if there's still anything unclear!

mikemahoney218 commented on R/di-fit.R:
```diff
@@ -0,0 +1,276 @@
+  dk <- calculate_dk(training, validation)
+
+  # 2.4 Dissimilarity index
+  d_bar <- mean(stats::dist(training))
```
Oh awesome -- I hadn't come across proxyC before. That'll simplify things a _lot_.

mikemahoney218 commented:
Thanks for the review! I believe I addressed most of your comments.

On the subject of variable importance:

* The new `tidy_importance` function is a thin generic that translates `vi` objects and `data.frame` objects into the broom-esque importance format we want. It's not currently exported, but not for any reason other than I don't know how useful it is.
* The new `check_di_importance` function will check to make sure that all variables have an importance estimate, all importance estimates have a variable, and `importance` is in our broom-esque format
* The documentation highlights that you need a 1:1 match between the variable importance table and the predictors in your data. It also specifies that you need to have dummy variables as separate predictors in your variable importance calculations.

That leaves this comment:
> Also, we need to differentiate between a zero importance due to scaling (like caret does) or due to not being used in the model. 

I don't know if this is solved by requiring all variables have importance estimates and all importance estimates have variables. If not, any chance you could elaborate on this?

mikemahoney218 commented on tests/testthat/_snaps/di-score.md:
```diff
@@ -0,0 +1,47 @@
+       7 0        TRUE 
+       8 1.15e- 9 TRUE 
+       9 2.04e-10 TRUE 
+      10 1.63e- 9 TRUE 
```
Sorry: I think I prefer returning small non-zero values to setting `digits` in `proxyC::dist()`, because I don't know how small a distance needs to be before it's non-meaningful.

mikemahoney218 commented:
Hey folks; I've got reasons to be suspicious of the implementation in this PR (I don't think it perfectly matches CAST, which is the implementation of record) and don't currently have time to investigate. I'm going to close this PR for now. 

I'm experimenting with this functionality in waywiser (which is why I found the problem with the math here), and if there's interest I can re-open a PR when I figure things out over there.
